keps/sig-ui/20190205-add-plugin-support-to-dashboard.md

```diff
@@ -0,0 +1,165 @@
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+  - [Proposal](#proposal)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+<!-- The `Summary` section is incredibly important for producing high quality user focused documentation such as release notes or a development road map.
+It should be possible to collect this information before implementation begins in order to avoid requiring implementors to split their attention between writing release notes and implementing the feature itself.
+KEP editors, SIG Docs, and SIG PM should help to ensure that the tone and content of the `Summary` section is useful for a wide audience.
+
+A good summary is probably at least a paragraph in length. -->
+
+We want to introduce a plugin mechanism in dashboard which shall enable easy functionality addition to the existing dashboard. 
+
+## Motivation
+
+<!-- This section is for explicitly listing the motivation, goals and non-goals of this KEP.
+Describe why the change is important and the benefits to users.
+The motivation section can optionally provide links to [experience reports][] to demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports -->
+
+The idea was proposed back in 2017 [here](https://github.com/kubernetes/dashboard/issues/1832). This mechanism will also be helpful in better utilisation of [CRDs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) through the dashboard.
+
+A custom UI requirement came up for [ingress-nginx](https://github.com/kubernetes/ingress-nginx/issues/109) which later required [enhancement](https://github.com/kubernetes/ingress-nginx/issues/2480) and dashboard should be able to support it without requiring constant changes to `kubernetes/dashboard`.
+
+In general, users should be able to add functionality without modifying the dashboard source code.
+
+### Goals
+
+<!-- List the specific goals of the KEP.
+How will we know that this has succeeded? -->
+
+- A 3rd party developer should be able to install a custom plugin to the dashboard extending its functionality in some way.
+
+<!-- ### Non-Goals -->
+
+<!-- What is out of scope for his KEP?
+Listing non-goals helps to focus discussion and make progress. -->
+
+## Proposal
+
+<!-- This is where we get down to the nitty gritty of what the proposal actually is. -->
+
+The plugin mechanism will comprise of largely of backend resources and frontend abstration that will help 3rd party developers to write plugins independently and register them with the dashboard.  
+
+Following are the major areas of focus:  
+#### How can a plugin be registered?  
+- Plugin Discovery  
+  - A new kind called `Plugin` should be introduced which MUST have a property specifying the location of the plugin bundle.
```

Comment:
Are there any requirements for the location? Will it be mounted inside Dashboard container? Would be good to add a few words of explanation here.

Reply:
Regarding where it should be mounted I'm not sure. I'm looking for suggestions on this. And yes, I think that plugin source can be fetched from a URL.

Reply:
Plugin Discovery would be nice if it was a CR and not require a mount. This would allow you to add plugins to the dashboard simply by helm installing the plugin. It would register its CR and stand up a web server sharing out the plugin bundle.

Reply:
We want to support different source types for plugin bundle `[URL|ConfigMap|Volume|Secret]`. It will be then retrieved from a specific location and served by our backend.

Reply:
If it has to be done by volume mount though, then it requires changing the deployment object used to maintain kube-dashboard. So, helm chart changes or whatever, then upgrading the existing dashboard to enable the plugin.
URL's totally external so shouldn't require changing the dashboard packaging at all. ConfigMap/Secret could also maybe be done this way as kube-dashboard could call into k8s to get them rather then volume mounting them.
If Volume were to be supported though, I can't see a way around needing to change the kube-dashboard packaging for each plugin enabled/disabled.